FV version of multiphase, multicomponent PorousFlow materials

cpgr opened the pull request:
This is the next part of making PorousFlow work with FV, and hopefully the hardest!

cpgr committed:
Grad(T) should be available for AD variables

The AD gradient of temperature should be availalbe for non-nodal
AD materials.

Refs #21275

cpgr committed:
Add AD thermophysical properties method

These UserObjects already use AD to compute all the derivatives,
but then passed all values and derivatives as Reals. For AD or FV
PorousFlow models, they should pass everything as ADReal.

Also lots of DualReal to ADReal for consistency.
Refs #21275

cpgr committed:
Template PorousFlowFluidState for AD

Refs #21275

cpgr committed:
Move common fluid state material code to a base class

This removes duplicate code in the fluid state materials.

Refs #21275

cpgr committed:
FluidState materials now inherit from base class

Refs #21275

cpgr committed:
Make interface to fluid state UO's consistent

Also fix up some inconsistencies in how these are formulated.

Refs #21275

cpgr committed:
Test ADPorousFlowFluidState with FV variables

Refs #21275

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1585473/) on d56ebd6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24668/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24668/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d3441a259ffa89fb85df106fa6b631f961ea7dad`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1585558/) on ed9ca62 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>d3441a</th>
      <th colspan=4>#24668 ed9ca6</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.73%</td>
      <td>95.70%</td>
      <td>-0.03%</td>
      <td>99.16%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10783</td>
      <td>10697</td>
      <td>-86</td>
      <td>588</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>481</td>
      <td>481</td>
      <td>-</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24668/coverage_diff/porous_flow)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24668/coverage/porous_flow)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24668/coverage/xfem)

</details>



This comment will be updated on new commits.

cpgr committed:
Template PorousFlowFluidMass Postprocessor

This Postprocessor will now work for AD and non-AD FE variables,
as well as FV variables.

Refs #21275

cpgr commented:
I think that with these changes I'll be able to add the example for #24590.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1588230/) on c021737 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24668/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24668/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4278fbb79d3acc746326873e9f53539d9aba697f`

GiudGiud commented:
I should be able to take a first look this weekend.

cpgr commented:
You can leave it until next week!

GiudGiud commented on modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:
```diff
@@ -0,0 +1,197 @@
+
+  // Set the initial values of the properties at the nodes.
+  // Note: not required for qp materials as no old values at the qps are requested
+  // unless the material is AD (for the FV case there are no nodal materials)
```
there is no AD old material properties though, the derivatives wouldnt match the current value of the solution

GiudGiud replied:
ok I saw they're not declared, just a nullptr

GiudGiud commented on modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:
```diff
@@ -0,0 +1,197 @@
+void
+PorousFlowFluidStateBaseMaterialTempl<is_ad>::setMaterialVectorSize() const
+{
+  _fluid_density[_qp].assign(_num_phases, 0.0);
```
since you're going to fill those, you dont need this assign, you could have resize everywhere instead of assign and save the "setting to 0" operation

GiudGiud replied:
kind of a micro-optimization though. Porbably not needed

cpgr replied:
I'll probably change it anyway

GiudGiud reviewed (commented):
quick look, skipped two large files

I think the handling of AD is mostly similar to what we did elsewhere.
I ll need to look more at the FV stuff

cpgr commented on modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:
```diff
@@ -0,0 +1,114 @@
+  /// Gradient of the mass fraction matrix (only defined at the qps)
+  GenericMaterialProperty<std::vector<std::vector<RealGradient>>, is_ad> * const _grad_mass_frac_qp;
+  /// Derivative of the mass fraction matrix with respect to the Porous Flow variables
+  MaterialProperty<std::vector<std::vector<std::vector<Real>>>> * const _dmass_frac_dvar;
```
Yeah, they are needed for the non-AD version, but are set to nullptr for the AD version

cpgr commented on modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:
```diff
@@ -0,0 +1,114 @@
+  using PorousFlowVariableBaseTempl<is_ad>::_num_pf_vars;
+};
+
+#define usingPorousFlowFluidStateBaseMaterialMembers                                               \
```
Do you mean define `usingFooMembers` and a shorter `usingADFooMembers`? Then is it just
```
if (is_ad)
   usingADFooMembers
else
  usingFooMembers
```

GiudGiud replied:
yeah for example!
I think you may want a constexpr in front of is_ad too

cpgr replied:
I tried this in the last commit. I couldn't use constexpr though, as it wouldn't compile?

cpgr commented on modules/porous_flow/include/userobjects/PorousFlowBrineCO2.h:
```diff
@@ -442,11 +449,11 @@ class PorousFlowBrineCO2 : public PorousFlowFluidStateMultiComponentBase
+                     const ADReal & temperature,
+                     const ADReal & co2_density,
+                     ADReal & A,
+                     ADReal & B) const;
```
I thought that ADReal and DualReal are the same now, and that ADReal was preferred?

cpgr commented on modules/porous_flow/src/userobjects/PorousFlowWaterVapor.C:
```diff
@@ -40,7 +40,7 @@ PorousFlowWaterVapor::PorousFlowWaterVapor(const InputParameters & parameters)
   _num_components = 1;
   _gas_phase_number = 1 - _aqueous_phase_number;
-  _gas_fluid_component = 1 - _aqueous_fluid_component;
+  _gas_fluid_component = _aqueous_fluid_component;
```
Yeah, I'm not sure what is best here. There is only one fluid component (water) that can exist in either the liquid or vapour phase. Maybe I should just use a single `_fluid_component` to make it clear? I could just use 0 but this makes it similar to the other, multiple component classes.

cpgr commented on modules/porous_flow/src/userobjects/PorousFlowWaterVapor.C:
```diff
@@ -121,73 +141,78 @@ PorousFlowWaterVapor::thermophysicalProperties(Real pressure,
+      liquid.pressure = pressure;
+      gas.pressure = pressure + _pc.capillaryPressure(liquid.saturation, qp);
+
+      gas.mass_fraction[_gas_fluid_component] = 1.0;
```
Yeah, not the vapour fraction (saturation).

cpgr commented on modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:
```diff
@@ -0,0 +1,197 @@
+                               : &this->template declareProperty<std::vector<std::vector<Real>>>(
+                                     "dPorousFlow_fluid_phase_density" + _ext + "_dvar")),
+    _fluid_viscosity(this->template declareGenericProperty<std::vector<Real>, is_ad>(
+        "PorousFlow_viscosity" + _ext)),
```
Hehehe - I've thought for years that I should delete the 'phase' from the material property name, as we don't use it for pressure, saturation, mass fraction etc.

GiudGiud commented on modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C:
```diff
@@ -126,261 +82,215 @@ PorousFlowFluidStateSingleComponent::PorousFlowFluidStateSingleComponent(
-            (*_dsaturation_dvar)[_qp][_aqueous_phase_number][_hvar] +
-        dpc * (*_dgrads_qp_dv)[_qp][_aqueous_phase_number][_hvar];
-  }
+    if constexpr (!is_ad)
```
could be an idea to propagate that to the other if (the use of constexpr)

GiudGiud replied:
why do you not need second derivatives with AD? AD doesnt do that for you does it?

GiudGiud commented on modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C:
```diff
@@ -126,261 +82,215 @@ PorousFlowFluidStateSingleComponent::PorousFlowFluidStateSingleComponent(
+    {
+      // Need to compute second derivatives of properties wrt variables for some of
+      // the gradient derivatives. Use finite differences for now
+      const Real dp = 1.0e-2;
```
better to make those relative to the local pressure (more resilient to regimes / units)
so 1e-5 * _liquid_porepressure[_qp] for example

GiudGiud commented on modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C:
```diff
@@ -126,261 +82,215 @@ PorousFlowFluidStateSingleComponent::PorousFlowFluidStateSingleComponent(
+
+      const auto d2T_dph = (fsp_dp[_aqueous_phase_number].temperature.derivatives()[_hidx] -
+                            _fsp[_aqueous_phase_number].temperature.derivatives()[_hidx]) /
+                               (2.0 * dp) +
```
why 2 dp ? you only added 1 inside the df(p + dp)/dT term

GiudGiud commented on modules/porous_flow/src/materials/PorousFlowFluidState.C:
```diff
@@ -141,148 +89,118 @@ PorousFlowFluidState::PorousFlowFluidState(const InputParameters & parameters)
-  Real Tk = _temperature[_qp] + _T_c2k;
+  const GenericReal<is_ad> Tk = _temperature[_qp] + _T_c2k;
 
+  _fs.clearFluidStateProperties(_fsp);
```
is the call right below not setting all the properties?

cpgr committed:
Revisions after code review

Refs #21275

cpgr committed:
FV-specific fluid mass postprocessor

Combining this with the existing templated FE-specific postprocessor
was a little clunky. Even though there is some duplication, it is much
cleaner to follow the logic.

Refs #21275

GiudGiud commented on modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md:
```diff
@@ -0,0 +1,28 @@
+\end{equation*}
+where all variables are defined in [`nomenclature`](/nomenclature.md).
+
+The fluid component $\kappa$ is specified in the input parameter `fluid_component`.
```
```suggestion
The fluid component $\kappa$ is specified in the input parameter [!param](/Postprocessors/FVPorousFlowFluidMass/fluid_component).
```

GiudGiud commented on modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md:
```diff
@@ -0,0 +1,28 @@
+The fluid component $\kappa$ is specified in the input parameter `fluid_component`.
+By default, the mass of fluid component $\kappa$ is summed over all fluid phases. The
+sum can be restricted to only a subset of fluid phases by entering the phase indexes
+in the `phase` input parameter.
```
```suggestion
in the [!param](/Postprocessors/FVPorousFlowFluidMass/phase) input parameter.
```

GiudGiud commented on modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md:
```diff
@@ -0,0 +1,28 @@
+in the `phase` input parameter.
+
+This `Postprocessor` also provides the option to only calculate fluid mass below a
+certain saturation, which can be invoked using the `saturation_threshold` parameter.
```
```suggestion
certain saturation, which can be invoked using the [!param](/Postprocessors/FVPorousFlowFluidMass/saturation_threshold) parameter.
```

GiudGiud commented on modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md:
```diff
@@ -0,0 +1,28 @@
+
+The flag `use_displaced_mesh = false` is set internally by this Postprocessor, and the parameter cannot be altered by the user, even for simulations with solid-mechanical deformation.  The reason is that this Postprocessor uses the strain calculated by TensorMechanics to automatically compensate for deformed meshes.  Further information may be found [here](porous_flow/time_derivative.md).  Theref ...
+
+- For mechanically-coupled simulations, you must provide a `base_name` that is identical to that used by the TensorMechanics strain calculator, so that this Postprocessor can retrieve the correct strain.  The most common use-case is that you provide no `base_name` to the TensorMechanics strain calculator and hence no `base_name` to this Postprocessor.
```
```suggestion
- For mechanically-coupled simulations, you must provide a [!param](/Postprocessors/FVPorousFlowFluidMass/base_name) parameter that is identical to that used by the TensorMechanics strain calculator, so that this Postprocessor can retrieve the correct strain.  The most common use-case is that you provide no `base_name` to the TensorMechanics strain calculator and hence no `base_name` to this Postprocessor.
```

GiudGiud replied:
once is enough to make sure it doesnt change silently breaking the docs imo

GiudGiud commented on modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:
```diff
@@ -0,0 +1,114 @@
+   * @param value AD value
+   * @return raw or AD value depending on is_ad
+   */
+  GenericReal<is_ad> genericValue(const ADReal & value);
```
so weird that we dont have that elsewhere
@lindsayad do we not?

GiudGiud replied:
if we dont, I could see this in MooseTypes.h.
what do you think?

lindsayad replied:
I would prefer not to. I generally don't like to add more APIs for dropping derivatives. But I would be open to changing my mind

lindsayad replied:
I'm fine with you having it in your module. And if other modules come up with justifiable uses then we could migrate it to the framework

GiudGiud commented on modules/porous_flow/include/postprocessors/FVPorousFlowFluidMass.h:
```diff
@@ -0,0 +1,50 @@
+#include "PorousFlowDictator.h"
+
+/**
+ * Postprocessor produces the mass of a given fluid component in a region.
```
```suggestion
 * Postprocessor that produces the mass of a given fluid component in a region.
```

GiudGiud commented on modules/porous_flow/src/postprocessors/FVPorousFlowFluidMass.C:
```diff
@@ -0,0 +1,126 @@
+    _dictator(getUserObject<PorousFlowDictator>("PorousFlowDictator")),
+    _fluid_component(getParam<unsigned int>("fluid_component")),
+    _phase_index(getParam<std::vector<unsigned int>>("phase")),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _has_total_strain(hasMaterialProperty<RankTwoTensor>(_base_name + "total_strain")),
```
this logic seems dangerous

if i mispelled base_name, it wont retrieve a total strain and will change the result.

why would not always retrieve the total strain if a base name is passed?

cpgr replied:
Oh, you might want to use a total strain but if you make a typo in the input file, it will just silently not use one and the user won't know why they are getting an incorrect value. I guess we should error if base_name is provided but doesn't actually exist.

This was a copy and paste from the existing post processor, and I think the same logic is used in the energy postprocessor too. I'll make an issue to fix all of these.

cpgr replied:
I've opened an issue to fix this in the postprocessors not included in this PR

cpgr replied:
I added a check and a paramError when base_name is given by doesn't exist. I think this is probably the correct behaviour.

GiudGiud reviewed (commented):
cursory review

cpgr committed:
Only include derivative members if not using AD

Refs #21275

cpgr committed:
Avoid duplicate functor error when running test

Refs #21275

cpgr committed:
Error when an incorrect strain base name is provided

Refs #25673

GiudGiud commented on modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C:
```diff
@@ -126,261 +77,215 @@ PorousFlowFluidStateSingleComponent::PorousFlowFluidStateSingleComponent(
-            (*_dsaturation_dvar)[_qp][_aqueous_phase_number][_hvar] +
-        dpc * (*_dgrads_qp_dv)[_qp][_aqueous_phase_number][_hvar];
-  }
+    if constexpr (!is_ad)
```
should this exclude nodal materials too?

cpgr replied:
Yep, it has the same logic as the comment above

GiudGiud reviewed (commented):
last set of comments! This looks good.

cpgr commented on modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:
```diff
@@ -0,0 +1,197 @@
+void
+PorousFlowFluidStateBaseMaterialTempl<is_ad>::setMaterialVectorSize() const
+{
+  _fluid_density[_qp].resize(_num_phases, 0.0);
```
They are sized in the call to
```
// Size pressure and saturation and prepare the derivative vectors
PorousFlowVariableBaseTempl<is_ad>::computeQpProperties();
```

(I added some inline documentation to make it clear)

cpgr commented on modules/porous_flow/src/materials/PorousFlowFluidState.C:
```diff
@@ -8,118 +8,66 @@
+    _Xnacl(_nodal_material && _is_Xnacl_nodal
+               ? this->template coupledGenericDofValue<is_ad>("xnacl")
+               : this->template coupledGenericValue<is_ad>("xnacl")),
+    _grad_Xnacl_qp(this->template coupledGenericGradient<is_ad>("xnacl")),
```
Yeah, we don't use them in that case

cpgr commented on modules/porous_flow/src/materials/PorousFlowFluidState.C:
```diff
@@ -316,117 +234,86 @@ PorousFlowFluidState::computeQpProperties()
 
     // Derivatives of gradients wrt variables
-    if (_dictator.isPorousFlowVariable(_gas_porepressure_varnum))
+    if constexpr (!is_ad)
```
Yep. I had to do this earlier though to remove it
```
  if (!_nodal_material)
    if constexpr (!is_ad)
    {
```
I couldn't get these into the one if statement

cpgr committed:
Remove extraneous if statement in PorousFlowFluidState

Refs #21275

GiudGiud commented:
looks great!

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1810521/) on b001fa5 : invalidated by @cpgr

cpgr commented:
Thanks for all your help @GiudGiud

GiudGiud commented:
my pleasure.
You have merge rights now right?

cpgr commented:
> my pleasure.
> You have merge rights now right?

I think I do, but is it ok to merge my own work as long as someone has reviewed it? And is there a way to mark all of the conversations as resolved?

GiudGiud commented:
There is no way to mark all conversations as resolved. You have to click them one by one.
Yes it's ok to merge your own work as long as you are the assignee and someone has approved it.
Maybe wait a little on that one, I have to check with Cody on something

cpgr commented on modules/porous_flow/src/postprocessors/PorousFlowFluidMass.C:
```diff
@@ -54,24 +56,29 @@ PorousFlowFluidMass::validParams()
   return params;
```
Ended up making a FV-specific one

GiudGiud commented:
ah yeah forgot I had assigned myself. We dont enforce that one strictly, it's more of a courtesy
